### ISSUE TITLE
[lex.digraph] p2 The requirement for alternative tokens should be relaxed for attributes

### ISSUE THREAD

**Issue by xmh0511** [cplusplus/CWG#266]
Full name of submitter (unless configured in github; will be published with the issue): Jim X

[lex.digraph] p2 says
> **In all respects of the language**, each alternative token **behaves the same, respectively, as its primary token**, except for its spelling. The set of alternative tokens is defined in Table 3.

However, things will become different if the alternative tokens that satisfy the syntactic requirements of an identifier appear in *attribute-token*, the same is true for [lex.key] p2
> Furthermore, the alternative representations shown in Table 6 for certain operators and punctuators ([lex.digraph]) are reserved and **shall not be used otherwise**.

### Suggested resolution

Change [lex.digraph] p2
> In all respects of the language, each alternative token behaves the same, respectively, as its primary token, except for its spelling or when it satisfies the syntactic requirements of an *identifier* and appears in an *attribute-token*. The set of alternative tokens is defined in Table 3. 

Change [lex.key] p2
> Furthermore, the alternative representations shown in Table 6 for certain operators and punctuators ([lex.digraph]) are reserved and shall not be used otherwise, except that the alternative tokens that satisfy the syntactic requirements of an *identifier* appear in *attribute-token*.

**Comment by villevoutilainen** [comment]
I don't see why this is reported as a Core issue. I expect 'and' and 'or' to mean the same thing as they mean elsewhere even inside attribute lists. If you have a reason to do otherwise, write a proposal and send it to EWG.

**Comment by jensmaurer** [comment]
Agreed. Unless we get an actual proposal for an attribute-token called `or` or `and`, there is no issue here.

**Comment by xmh0511** [comment]
@jensmaurer @villevoutilainen  Hmm, I didn't try to change the meaning here. I just said the mentioned wording are not consistent with [dcl.attr.grammar] p4
>  If a keyword or an alternative token that satisfies the syntactic requirements of an identifier ([lex.name]) is contained in an attribute-token, it is considered an identifier.

So, we shouldn't say: In all respects of the language, each alternative token behaves the same, respectively, as its primary token. Since *and* is that identifier, if we say "in all respects", that means, it will perform like *&&* in the *attribute-token*, which is not what  [dcl.attr.grammar] p4 meant.

**Comment by xmh0511** [comment]
> It's fine if `or` is treated as `||` in an attribute-token: That's ill-formed, and life continues from there.

Consider this example:
````cpp
int main(){
  int a [[or]];
}
````
All [implementations](https://godbolt.org/z/ePGoxjeKb) consider `or` is an *identifier*, which obeys what [dcl.attr.grammar] p4 said.

**Comment by jensmaurer** [comment]
> All [implementations](https://godbolt.org/z/ePGoxjeKb) consider or is an identifier, which obeys what [dcl.attr.grammar] p4 said.

That's inaccurate.  ICC says "error: expected an identifier", which seems to be what the standard prescribes ("||" is not an identifier).

> Why do we not give exceptions for alternative tokens that satisfy the requirements?

Because that would be normative duplication. We already say digraphs act the same as their corresponding tokens, thus when we get to [dcl.init.attr], we feed a non-identifier into the attribute grammar and fail.

Feel free to post bug reports to the compilers that don't give the right diagnostic. However, I'd like to point out that this is not a conformance concern, because quality of diagnostics is not prescribed in the standard.

**Comment by xmh0511** [comment]
> We already say digraphs act the same as their corresponding tokens, thus when we get to [dcl.init.attr], we feed a non-identifier into the attribute grammar and fail.

If it were that, [dcl.attr.grammar] p4 should not mention "alternative tokens" because all "alternative tokens" are always viewed as punctuators.

**Comment by jensmaurer** [comment]
Agreed.

**Comment by jensmaurer** [comment]
CWG2726
